dali/base/dautils.cpp

```diff
@@ -3785,3 +3785,103 @@ void logNullUser(IUserDescriptor * userDesc)
     }
 }
 #endif
+
+// FileReadPropertiesUpdater:
+// - aggregates the readCost and the numDiskReads for each logical file (using addCostAndNumReads method).
+// - upon publishing (calling publish method), it:
+//   - it traverses up the owning superfiles, computing the cumulative readCost and numDiskReads for the subfile owners.
+//   - it subsequently updates file properties for both the files and their owning superfiles with the newly derived statistics.
+// It is designed to optimize the update process for readCost and read count by:
+// * reducing the frequency of getOwningSuperFiles calls, as these operations are resource-intensive.
+// * performing a single-pass update of file properties (e.g. at the end of a graph execution).
+class FileReadPropertiesUpdater : public CSimpleInterfaceOf<IFileReadPropertiesUpdater>
+{
+    struct FileStatItem
+    {
+        Owned<IDistributedFile> file;
+        stat_type numDiskReads = 0;
+        cost_type readCost = 0;
+    };
+    std::unordered_map<std::string, FileStatItem> stats;
+    Linked<IUserDescriptor> udesc;
+
+    // Add the readCost and numDiskReads to owners stats
+    FileStatItem & appendOwnersCostAndNumReads(std::unordered_map<std::string, FileStatItem> & ownerStats, IDistributedFile * file, stat_type numDiskReads, cost_type curReadCost)
+    {
+        FileStatItem & curStatItem = ownerStats[file->queryLogicalName()];
+        if (!curStatItem.file)
+            curStatItem.file.setown(LINK(file));
+        curStatItem.numDiskReads += numDiskReads;
+        curStatItem.readCost += curReadCost;
+        return curStatItem;
+    }
+    // Update the readCost and numDiskReads for owning superfiles
+    // - this will recurse up the superfiles tree updating the owning superfiles
+    // n.b. fileStatItem.file must be set to a valid IDistributedFile
+    void updateOwnersStats(std::unordered_map<std::string, FileStatItem> & ownerStats, const FileStatItem & fileStatItem)
+    {
+        // getOwningSuperFiles iterator is slow, so only call it once (say at the end of graph)
+        Owned<IDistributedSuperFileIterator> iter = fileStatItem.file->getOwningSuperFiles();
+        ForEach(*iter)
+        {
+            IDistributedSuperFile & cur = iter->query();
```

Comment:
@jakesmith related to your comment above.  I wonder if there is scope for adding a flag to indicate it is ok to cache lookups (in a non-owning cache withing CDistributedFileDirectory).  In this situation it would mean the super file lookups for the other subfiles would be much more efficient (because the superfile will still have a reference).

Reply:
A cache of some sorts would be good. I've opened a linked JIRA to revisit: https://hpccsystems.atlassian.net/browse/HPCC-34172